Assets/MixedRealityToolkit.SDK/Experimental/Features/Input/Handlers/PrimitiveManipulator.cs

```diff
@@ -0,0 +1,744 @@
+            {
+                // Check pointer has a valid controller (e.g. gaze pointer doesn't)
+                if (p.pointer.Controller != null)
+                {
+                    numControllers++;
+                    sum += p.pointer.Controller.Velocity;
+                }
+            }
+            return sum / Math.Max(1, numControllers);
+        }
+
+        private Vector3 GetPointersAngularVelocity()
+        {
+            Vector3 sum = Vector3.zero;
+            int numControllers = 0;
+            foreach (var p in pointerIdToPointerMap.Values)
+            {
+                // Check pointer has a valid controller (e.g. gaze pointer doesn't)
+                if (p.pointer.Controller != null)
+                {
+                    numControllers++;
+                    sum += p.pointer.Controller.AngularVelocity;
+                }
+            }
+            return sum / Math.Max(1, numControllers);
+        }
+
+        private bool IsNearManipulation()
+        {
+            foreach (var item in pointerIdToPointerMap)
+            {
+                if (item.Value.IsNearPointer)
+                {
+                    return true;
+                }
+            }
+            return false;
+        }
+        #endregion Private Methods
+
+        #region Public Methods
+
+        /// <summary>
+        /// Releases the object that is currently manipulated
+        /// </summary>
+        public void ForceEndManipulation()
+        {
+            // end manipulation
+            if (isManipulationStarted)
+            {
+                HandleManipulationEnded(GetPointersGrabPoint(), GetPointersVelocity(), GetPointersAngularVelocity());
+            }
+            pointerIdToPointerMap.Clear();
+        }
+
+        /// <summary>
+        /// Gets the grab point for the given pointer id.
+        /// Only use if you know that your given pointer id corresponds to a pointer that has grabbed
+        /// this component.
+        /// </summary>
+        public Vector3 GetPointerGrabPoint(uint pointerId)
+        {
+            Assert.IsTrue(pointerIdToPointerMap.ContainsKey(pointerId));
+            return pointerIdToPointerMap[pointerId].GrabPoint;
+        }
+
+        #endregion Public Methods
+
+        #region Hand Event Handlers
+
+        /// <inheritdoc />
+        public void OnPointerDown(MixedRealityPointerEventData eventData)
+        {
+            if (!allowFarManipulation && eventData.Pointer as IMixedRealityNearPointer == null)
+            {
+                return;
+            }
+
+            // If we only allow one handed manipulations, check there is no hand interacting yet. 
+            if (manipulationType != HandMovementType.OneHanded || pointerIdToPointerMap.Count == 0)
+            {
+                uint id = eventData.Pointer.PointerId;
+                // Ignore poke pointer events
+                if (!eventData.used && !pointerIdToPointerMap.ContainsKey(id))
```

Comment:
>eventData.used [](start = 21, length = 14)

I think this is also an existing issue, but consider moving this !eventData.used check to the if statement at the beginning of this function (i.e. we don't do anything interesting if the eventData is used - I guess we MIGHT re-mark it as used, but since it's already used, that's also a no-op).